Non-root  scrolling

cfraz89 opened the pull request:
Implements scrolling on mouse-wheel events, of any node which is hovered. When a node has hit its scroll boundary, scrolling bubbles up to the node's parent, all the way up to the root element. The root (html) element is sized to the viewport (is this technically the right thing to do?) and therefore will automatically handle root-level scrolling.

Adds an example `scrolling` which demos scrolling in the middle element. Scrolling the root element can be tested by resizing the window smaller than the content and scrolling, otherwise the tailwind example has a large content area and should be scrollable too.

Todo: Touch scrolling, flinging with momentum (even with mouse wheel? chromium does this).

jkelleyrtp commented:
Thanks for the PR! Nico's currently on break so I'll take a look at it, but it'll take a moment :)

cfraz89 commented:
No problems

nicoburns commented on packages/blitz/src/renderer/render.rs:
```diff
@@ -354,12 +349,26 @@ impl<'dom> VelloSceneGenerator<'dom> {
         cx.stroke_border(scene);
         cx.stroke_devtools(scene);
+
+        //Now that background has been drawn, offset pos and cx in order to draw our contents scrolled
```
Nit: Comment formatting (please add space between `//` and `Now`)

nicoburns commented on packages/dom/src/node.rs:
```diff
@@ -797,6 +800,50 @@ impl Node {
+            || self.scroll_offset.y > self.final_layout.scroll_height() as f64
+        {
+            if let Some(parent) = self.parent {
+                unsafe {
```
If you moved the scroll_by method to the `Document` then you could avoid the `unsafe` here.

nicoburns commented on packages/dom/src/node.rs:
```diff
@@ -797,6 +800,50 @@ impl Node {
+        self.scroll_offset.y += y;
+
+        //If we're past our scroll bounds, transfer scrolling to parent
+        if self.scroll_offset.x < 0.0
+            || self.scroll_offset.x > self.final_layout.scroll_width() as f64
+            || self.scroll_offset.y < 0.0
+            || self.scroll_offset.y > self.final_layout.scroll_height() as f64
```
This check should occur before updating the node's scroll offset. It should also allow *partial* transfer of the scroll delta to the parent. For example, if `scroll_offset.y` is 10 and x is `-20` then the current node should scroll by `-10` to scroll offset `0`, and the parent should be scrolled by `10`.

nicoburns replied:
I believe if you did this then you wouldn't need to clamp.

nicoburns commented on packages/dom/src/document.rs:
```diff
@@ -739,31 +744,6 @@ impl Document {
     }
 
-    pub fn scroll_by(&mut self, px: f64) {
-        // Invert scrolling on macos
-        #[cfg(target_os = "macos")]
-        {
-            self.scroll_offset += px;
-        }
-        #[cfg(not(target_os = "macos"))]
-        {
-            self.scroll_offset -= px;
-        }
```
I think we still need this. This branch is scrolling in the incorrect direction for me. I am unsure if the inversion is macOS specific or scrolling needs to be inverted on all platforms.

cfraz89 replied:
Ok, I realised it was scrolling the wrong way on my gnome desktop too, so in the updated scrolling code I've inverted it. Could you double check it's scrolling the right way on mac now?

nicoburns commented on packages/dom/src/document.rs:
```diff
@@ -697,12 +692,22 @@ impl Document {
+        taffy::round_layout(self, root_element_id);
+
+        // Root element's size is viewport size
+        let root_element = self.get_node_mut(self.root_element().id).unwrap();
+
+        root_element.final_layout.size = taffy::Size {
+            width: size.width.to_f32_px(),
+            height: size.height.to_f32_px(),
+        };
+
+        root_element.clamp_scroll_offset();
```
I believe that the way the web works is that there is an implicit "viewport" node that lives above the HTML element that has a fixed size, with the HTML element itself being a normal node. This is probably close enough for the time being, although I don't like overriding the layout size (how does this work when generating "full height" screenshots?)

nicoburns replied:
It would be good to have a way to leave the viewport size undefined (determined by layout) as this is used to render "full height screenshots" in the screenshot example (compare output of `cargo run --release --example screenshot https://servo.org` on this branch vs. main)

cfraz89 replied:
Ok I've kind of restored the old behaviour where we keep a scroll property on the document, which gets bubbled up to from the root node. That seems to be the simplest way to avoid this issue for now, as I don't think we need to go so far as to actually have a viewport node in the tree yet.

nicoburns commented on packages/dioxus-blitz/src/window.rs:
```diff
@@ -409,6 +409,7 @@ impl<Doc: DocumentLike> View<Doc> {
             WindowEvent::Focused(_) => {}
 
             // Touch and motion events
+            //Todo implement touch scrolling
```
Nit: Comment formatting (please add space between `//` and `Todo`)

nicoburns reviewed (changes requested):
Thanks for the PR! Overall this looks like the correct approach, but I have few review comments

(comments prefixed with `Nit` are minor style issues that I would prefer to be addressed but will not block merging)

cfraz89 commented:
Thanks for the feedback, all makes sense. I revamped the PR:

- Scrolling behaviour is all implemented in document now
- Implemented the partial bubbling of scroll up to parent node, then to document scroll
- Removed locking root element size to viewport size, and instead track scroll on document again
- Flipped scroll direction 
- use kurbo::Point instead of taffy::Point to store scroll coords, realised that taffy::Point has a bit of a css slant
- Fixed comment formatting

cfraz89 commented on packages/dioxus-blitz/src/window.rs:
```diff
@@ -191,7 +191,7 @@ impl<Doc: DocumentLike> View<Doc> {
     pub fn mouse_move(&mut self, x: f32, y: f32) -> bool {
         let dom_x = x / self.viewport.zoom();
-        let dom_y = (y - self.dom.as_ref().scroll_offset as f32) / self.viewport.zoom();
+        let dom_y = y / self.viewport.zoom();
```
I see, have addressed this now, updated the scroll example to reflect it as well.
